NDTensors/ext/NDTensorsAMDGPUExt/adapt.jl

```diff
@@ -0,0 +1,29 @@
+using NDTensors: NDTensors, EmptyStorage, adapt_storagetype, emptytype
+using NDTensors.AMDGPUExtensions: AMDGPUExtensions, ROCArrayAdaptor
+using NDTensors.GPUArraysCoreExtensions: storagemode
+using NDTensors.TypeParameterAccessors:
+  set_type_parameter, set_type_parameters, type_parameter, type_parameters
+using Adapt: Adapt, adapt
+using AMDGPU: AMDGPU, ROCArray, ROCVector
+
+function AMDGPUExtensions.roc(xs)
+  return fmap(x -> adapt(ROCArrayAdaptor{AMDGPU.Runtime.Mem.HIPBuffer}(), x), xs)
```

Comment:
Maybe this should use `default_type_parameter(ROCArray, storagemode)`.

Reply:
In the other instances of this function we have the ability to modify the storagemode type so I am rewriting this as 
```
function AMDGPUExtensions.roc(xs, buffer=default_type_parameter(ROCArray, storagemode))
  return fmap(x -> adapt(ROCArrayAdaptor{buffer}(), x), xs)
end
```

Reply:
Should we call the keyword argument `storagemode` or `buffer`? What does `AMDGPU` call that?

Reply:
@mtfishman in AMDGPU they call it a buffer 
```
function ROCArray{T, N, B}(
        ::UndefInitializer, dims::Dims{N},
    ) where {T, N, B <: Union{Mem.HIPBuffer, Mem.HostBuffer}}
        @assert isbitstype(T) "ROCArray only supports bits types"
        buf = B(prod(dims) * sizeof(T); stream=stream())
        xs = new{T, N, B}(DataRef(_free_buf, buf), dims, 0)
        finalizer(unsafe_finalize!, xs)
        return xs
    end
    ```

Reply:
Right, but `CUDA.jl` also refers to it internally as a "buffer", but we've decided to use `storagemode` as a GPU-agnostic term for that, and the `AMDGPU.roc` function doesn't have a name for it: https://github.com/JuliaGPU/AMDGPU.jl/blob/v0.8.11/src/array.jl#L238. So perhaps we should go ahead and make our own decision to call that `storagemode` for any of the GPU backends.

Reply:
I can definitely change this name of this variable to `storagemode`.
Just a question in NDTensorsCUDAExt this function looks like this
```
function CUDAExtensions.cu(xs; unified::Bool=false)
  return fmap(
    x -> adapt(CuArrayAdaptor{unified ? Mem.UnifiedBuffer : Mem.DeviceBuffer}(), x), xs
  )
end
```
I know we discussed changing this to accept the CUDA Mem types but I think that was on a PR that was rewritten. So I can make an entry in our googledoc to change this in the next PR when I update the GPU backend files. And when I do I will change the variable to `storagemode`

Reply:
The boolean system was something they previously had implemented until they created the `Mem.HostBuffer` type. But I have made a note of this and will work on the other GPUs in the next PR. For now, this code is updated to be `storagemode`